keymapper/migrations.py

```diff
@@ -0,0 +1,118 @@
+from keymapper.paths import get_preset_path, mkdir, CONFIG_PATH
+
+
+def all_presets():
+    p = Path(get_preset_path())
+    presets = []
+    for folder in p.iterdir():
+        if folder.is_dir():
+            for preset in folder.iterdir():
+                if preset.suffix == ".json":
+                    presets.append(preset)
+    return presets
+
+def version():
+    config_path = os.path.join(CONFIG_PATH, "config.json")
+    config = {}
+    
+    if not os.path.exists(config_path):
+        return 0
+    with open(config_path, "r") as file:
+        config = json.load(file)
+    if "migration" in config.keys():
+        return config["migration"]
+    return 1
+
+def _config_suffix():
+    deprecated_path = os.path.join(CONFIG_PATH, "config")
+    config_path = os.path.join(CONFIG_PATH, "config.json")
+    if os.path.exists(deprecated_path) and not os.path.exists(config_path):
+        logger.info('Moving "%s" to "%s"', deprecated_path, config_path)
+        os.rename(deprecated_path, config_path)
+
+def _preset_path():
+    """Migrate the folder structure from < 0.4.0.
+
+    Move existing presets into the new subfolder "presets"
+    """
+    new_preset_folder = os.path.join(CONFIG_PATH, "presets")
+    if not os.path.exists(get_preset_path()) and os.path.exists(CONFIG_PATH):
+        logger.info("Migrating presets from < 0.4.0...")
+        groups = os.listdir(CONFIG_PATH)
+        mkdir(get_preset_path())
+        for group in groups:
+            path = os.path.join(CONFIG_PATH, group)
+            if os.path.isdir(path):
+                target = path.replace(CONFIG_PATH, new_preset_folder)
+                logger.info('Moving "%s" to "%s"', path, target)
+                os.rename(path, target)
+        logger.info("done")
+
+
+def _mapping_keys():
+    if not os.path.exists(get_preset_path()):
+        return # don't execute if there are no presets
+    for preset in all_presets():
+        preset_dict = {}
+        with open(preset, "r") as file:
+            preset_dict = json.load(file)
+        if "mapping" in preset_dict.keys():
+            mapping = copy.deepcopy(preset_dict["mapping"])
+            for key in mapping.keys():
+                if key.count(",") == 1:
+                    preset_dict["mapping"][key + ",1"] = preset_dict["mapping"].pop(key)
+        with open(preset, "w") as file:
+            json.dump(preset_dict, file, indent=4)
+            file.write("\n")
+
+def _update_version(version):
+    config_path = os.path.join(CONFIG_PATH, "config.json")
+    if not os.path.exists(config_path):
+        return
+    with open(config_path, "r") as file:
+        config = json.load(file)
+    config["migration"] = version
+    with open(config_path, "w") as file:
+        json.dump(config, file, indent=4)
+        
+def migrate():
+    v = version()
+    if v == 0:
+        _config_suffix()
+        v = 1
+    if v == 1:
+        _preset_path()
+        _mapping_keys()
+        v = 2
+        _update_version(2)
```

Comment:
would it make sense to use a semver string here?

`version.parse(v) < version.parse("1.0.0")`

https://stackoverflow.com/a/11887885/4417769

We would still store the version in the config to remember up to which point things were migrated. We just use a different format for it to see with which version this migration corresponds, to improve readability. So that we, as developers, know when this migration was supposed to run.

Reply:
Also, if someone makes a clean install, what will v be initialized with?

If we use a semver string here, we can place the current version at installation time into `INITIAL_CONFIG`

Reply:
makes totally sense. For a clean install we should not need any migrations. So probably initialize with the current version.